### PR TITLE
Limit the number of elements to build during prerender

### PR DESCRIPTION
**Pull request by cramforce**
This significantly reduces the amount of CPU needed for very large documents during prerender. The count of elements to build was chosen arbitray as "big enough that it typically should be enough to render stuff in the first viewport".

This seems to work quite well and it suggests we do have quite the opportunity to optimize things with a chunking scheduler.

The document I was testing with (http://www.refinery29.com/amp/best-pizza-nyc) does get much better with this, although it doesn't get into good shape, since it is simply too big for a browser (let alone a mobile browser) to reasonably render.

### CODE REVIEW

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -474,6 +477,10 @@ export class Resources {
     this.resources_.push(resource);
   }
 
+  shouldBuildNow() {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: jsdoc

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -474,6 +477,10 @@ export class Resources {
...
+    return this.buildAttemptsCount_++ < 20 || this.viewer_.hasBeenVisible();
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: The method name seems idempotent. How about something like `grantBuild`?

**Comment by cramforce** [comment]
Thanks for approval :)

This was meant as a real RFC. Do you think this is the right way to go?

**Comment by dvoytenko** [comment]
@cramforce As a short path this would work. Provided you confirmed that the elements will be picked up later when viewer becomes visible, which I believe happens now in passes.

Ideally, though, we'd be a bit more selective and ensure that we do not run too many passes over prerendered document, constantly trying to rebuild.

**Comment by cramforce** [comment]
I've never seen these passes to be an issue. We do always schedule a pass when we become visible.

Getting data from @lannka on how many `amp-` tags are on typical pages.

File: src/service/resource.js [hunk 3]
```diff
@@ -282,7 +282,8 @@ export class Resource {
    * for details.
    */
   build() {
-    if (this.blacklisted_ || !this.element.isUpgraded()) {
+    if (this.blacklisted_ || !this.element.isUpgraded()
+        || !this.resources_.shouldBuildNow()) {
```
Comment (jridgewell) [inline src/service/resource.js]: This check can be moved up into `Resources#buildOrScheduleBuildForResource_`

**Comment by lannka** [comment]
@cramforce on average 10 AMP elements. 1.2 `amp-ad`.

**Comment by cramforce** [comment]
Any chance you could get 95th and 99th percentile?

File: src/service/resources-impl.js (hunk 1 above)
Comment (cramforce) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js (hunk 2 above)
Comment (cramforce) [inline src/service/resources-impl.js]: Changed to `grantBuildPermission`

File: src/service/resource.js (hunk 3 above)
Comment (cramforce) [inline src/service/resource.js]: Unfortunately there is a second callsite.

**Comment by cramforce** [comment]
Friendly ping :)

File: src/service/resource.js [hunk 4]
```diff
@@ -285,6 +285,9 @@ export class Resource {
     if (this.blacklisted_ || !this.element.isUpgraded()) {
       return;
     }
+    if (!this.resources_.grantBuildPermission()) {
+
```
Comment (jridgewell) [inline src/service/resource.js]: Um?

Comment (dvoytenko) [inline src/service/resource.js]: Should there be something in this `if`?

Comment (cramforce) [inline src/service/resource.js]: Wow, how did this pass the tests? :)

Comment (cramforce) [inline src/service/resource.js]: So, weird. The local test didn't pass. Anyway, made the test more robust.

**Pull request merged**